src/emulator/functionsEmulatorRuntime.ts

```diff
@@ -498,99 +497,142 @@ async function getGRPCInsecureCredential(frb: FunctionsRuntimeBundle): Promise<a
     })
-    .when("firestore", (adminModuleTarget) => {
-      // When we call admin.firestore() we want to forward this on to the default app,
-      // but if you access something like admin.firestore.FieldValue we don't want to
-      // do anything.
-      const adminFirestoreProxy = new Proxied<typeof admin.firestore>(
-        adminModuleTarget.firestore
-      ).applied(() => {
-        return defaultApp.firestore();
-      });
-
-      return adminFirestoreProxy.finalize();
+    .when("firestore", (target) => {
+      if (frb.ports.firestore) {
+        return proxiedFirestore;
+      } else {
+        warnAboutFirestoreProd();
+        return Proxied.getOriginal(target, "firestore");
+      }
+    })
+    .when("database", (target) => {
+      if (frb.ports.database) {
+        return proxiedDatabase;
+      } else {
+        warnAboutDatabaseProd();
+        return Proxied.getOriginal(target, "database");
+      }
     })
     .finalize();
 
   // Stub the admin module in the require cache
   require.cache[adminResolution.resolution] = {
-    exports: adminModuleProxy,
+    exports: proxiedAdminModule,
   };
 
   logDebug("firebase-admin has been stubbed.", {
     adminResolution,
   });
-
-  return adminModuleProxy;
 }
 
-function proxyFirebaseApp(
+function makeProxiedFirebaseApp(
   frb: FunctionsRuntimeBundle,
-  original: admin.app.App,
-  sslCreds: any
+  original: admin.app.App
 ): admin.app.App {
-  const initializeFirestoreSettings = (firestoreTarget: any, userSettings: any) => {
-    const isEnabled = isFeatureEnabled(frb, "admin_stubs");
-
-    if (!isEnabled) {
-      if (!hasInitializedFirestore) {
-        firestoreTarget.settings(userSettings);
-        hasInitializedFirestore = true;
+  const appProxy = new Proxied<admin.app.App>(original);
+  return appProxy
+    .when("firestore", (target: any) => {
+      if (frb.ports.firestore) {
+        return proxiedFirestore;
+      } else {
+        warnAboutFirestoreProd();
+        return Proxied.getOriginal(target, "firestore");
       }
-      return;
-    }
+    })
+    .when("database", (target: any) => {
+      if (frb.ports.database) {
+        return proxiedDatabase;
+      } else {
+        warnAboutDatabaseProd();
+        return Proxied.getOriginal(target, "database");
+      }
+    })
+    .finalize();
+}
 
+function makeProxiedDatabase(target: typeof admin): typeof admin.database {
+  return new Proxied<typeof admin.database>(target.database)
+    .applied(() => {
+      return databaseApp.database();
+    })
+    .finalize();
+}
+
+async function makeProxiedFirestore(
+  frb: FunctionsRuntimeBundle,
+  target: any
+): Promise<typeof admin.firestore> {
+  // If we can't get sslCreds that means either grpc or grpc-js doesn't exist. If this is the save,
+  // then there's probably something really wrong (like a failed node-gyp build). If that's the case
+  // we should silently fail here and allow the error to raise in user-code so they can debug appropriately.
```

Comment:
nit: `silent fail` => `swallow error`? I think that may explain the idea a little better.